galsim/real.py

```diff
@@ -839,3 +905,426 @@ def _parse_files_dirs(file_name, image_dir, dir, noise_dir, sample):
+            # xscale = 2*np.pi/(nx*img.scale)
+            # yscale = 2*np.pi/(ny*img.scale)
+            # tmpwcs = galsim.JacobianWCS(yscale, 0.0, 0.0, xscale)
+            # tmp = galsim.ImageC(pk+0j, wcs=tmpwcs)
+            # iki = galsim.InterpolatedKImage(tmp)
+            # pk = iki.drawKImage(nx=nkx, ny=nky, wcs=wcs)
+            # pks[i] = pk.array.real * nx * ny
+
+        w = 1./np.sqrt(pks)
+
+        # Allocate and fill output coefficients and covariances.
+        # Note: put NSED axis last, since significantly faster to compute them this way,
+        # even though we eventually convert to images which are strided in this format.
+        coef = np.zeros((nky, nkx, NSED), dtype=np.complex128)
+        Sigma = np.empty((nky, nkx, NSED, NSED), dtype=np.complex128)
+
+        # Solve the weighted linear least squares problem for each Fourier mode.  This is
+        # effectively a constrained chromatic deconvolution.  Take advantage of symmetries.
+        if False:  # MJ: Leaving this here for now.  Also repeated in RealGalaxy.cpp, so can
+                   #     probably remove it.  But I'll let Josh do that.
+                   # Note: On my machine, with this branch __init__ takes 73 sec when running
+                   #       through test_real.py.  With the other branch, it takes 7.3 sec.
+                   #       So a full order of magnitude speed up doing this in C++ with TMV.
+          for ix in range(nkx//2+1):
+            for iy in range(nky):
+                if (ix == 0 or ix == nkx//2) and iy > nky//2:
+                    break # already filled in the rest of this column
+                ww = np.diag(w[:, iy, ix])
+                A = np.dot(ww, PSF_eff_kimgs[:, :, iy, ix])
+                b = np.dot(ww, kimgs[:, iy, ix])
+                try:
+                    x, resids, rank, singval = np.linalg.lstsq(A, b)
+                    # condition number is max singular value over min singular value
+                    condnum = np.max(singval) / np.min(singval)
+                    # Only bother computing covariance of result if condition number is favorable.
+                    if condnum < 1.e12:
+                        dx = np.linalg.inv(np.dot(np.conj(A.T), A))
+                    else:
+                        dx = np.zeros((NSED, NSED), dtype=np.complex128)
+                except Exception as e:
+                    x = 0.0
+                    dx = np.zeros((NSED, NSED), dtype=np.complex128)
+                coef[iy, ix] = x
+                Sigma[iy, ix] = dx
+                # Save work by filling in conjugates.
+                coef[-iy, -ix] = np.conj(x)
+                Sigma[-iy, -ix] = np.conj(dx)
+        else:
+          galsim._galsim.ComputeCRGCoefficients(
+            coef.ctypes.data, Sigma.ctypes.data,
+            w.ctypes.data, kimgs.ctypes.data, PSF_eff_kimgs.ctypes.data,
+            NSED, Nim, nkx, nky)
```

Comment:
@jmeyers314 FYI, I re-wrote this section in C++, which sped it up by an order of magnitude.  So np.lstsq is no longer the tall pole for running the unit tests at least.  It's mostly now drawImage and drawKImage.